controllers/observability/perses.go

```diff
@@ -0,0 +1,227 @@
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+)
+
+// persesAssetsDir is the directory (inside the repo image) containing Perses CR manifests (yaml).
+// Each YAML should define a single resource: PersesDashboard or PersesDatasource.
+const persesAssetsDir = "assets/dashboards/perses"
+
+// ReconcilePersesResources applies all Perses resources shipped with the operator under assets/perses.
+// It uses server-side apply to create/update resources idempotently without relying on cache reads.
+func (r *Reconciler) ReconcilePersesResources(ctx context.Context) error {
+	wd, err := os.Getwd()
+	if err != nil {
+		return fmt.Errorf("failed to get working dir: %w", err)
+	}
+
+	d := os.DirFS(wd)
+	// If the directory does not exist, nothing to do.
+	if _, err := fs.Stat(d, persesAssetsDir); err != nil {
+		return nil
+	}
+
+	// Ensure Perses CRDs are present before attempting to apply resources.
+	// Without CRDs, server-side apply will fail on every reconcile loop.
+	if !r.persesCRDsReady(ctx) {
+		return nil
+	}
+
+	// Load and cache assets once
+	var loadErr error
+	r.assetsOnce.Do(func() {
+		r.cachedAssets, loadErr = readPersesAssets(d)
+	})
+	if loadErr != nil {
+		return loadErr
+	}
+
+	// Apply cached assets each reconcile
+	for _, raw := range r.cachedAssets {
+		objMap := deepCopyMap(raw)
+		// enforce namespace
+		metadata, _ := objMap["metadata"].(map[string]any)
+		if metadata == nil {
+			metadata = map[string]any{}
+			objMap["metadata"] = metadata
+		}
+		metadata["namespace"] = r.namespace
+
+		apiVersion, _ := objMap["apiVersion"].(string)
+		kind, _ := objMap["kind"].(string)
+		if apiVersion == "" || kind == "" {
+			return fmt.Errorf("missing apiVersion/kind in cached asset")
+		}
+		u := &unstructured.Unstructured{Object: objMap}
+		u.SetGroupVersionKind(schema.FromAPIVersionAndKind(apiVersion, kind))
+
+		if err := r.Patch(ctx, u, client.Apply, client.FieldOwner("hyperconverged-operator"), client.ForceOwnership); err != nil {
+			return fmt.Errorf("failed to apply %s/%s: %w", u.GetNamespace(), u.GetName(), err)
+		}
+	}
+	return nil
+}
+
+// persesCRDsReady returns true when Perses CRDs exist in the cluster.
+func (r *Reconciler) persesCRDsReady(ctx context.Context) bool {
+	crdGVK := schema.GroupVersionKind{
+		Group:   "apiextensions.k8s.io",
+		Version: "v1",
+		Kind:    "CustomResourceDefinition",
+	}
+	for _, crdName := range []string{
+		"persesdashboards.perses.dev",
+		"persesdatasources.perses.dev",
+	} {
+		u := &unstructured.Unstructured{}
+		u.SetGroupVersionKind(crdGVK)
+		if err := r.Get(ctx, types.NamespacedName{Name: crdName}, u); err != nil {
+			return false
+		}
+	}
+	return true
+}
+
+// readPersesAssets walks the assets dir and returns parsed objects as generic maps.
+func readPersesAssets(d fs.FS) ([]map[string]any, error) {
+	var out []map[string]any
+	err := fs.WalkDir(d, persesAssetsDir, func(path string, entry fs.DirEntry, err error) error {
```

Comment:
Reading from disk is expensive, and doing it again and again for files that will never changed, is not so efficient. This whole function must be called only once, on boot time: Collect the files into objects in the memory, and then use these object for reconcile.

 See the implementation of quickstart, or imagestream for example.